File: ch6/play_again2.c
// set tty into char-by-char mode and no-echo mode,read char, return result

#include <stdio.h>
#include <termios.h>

#define QUESTION "Do you want another translation"

int get_response(char * question){

    int input;
    printf("%s (y/n)?",question);
    while (1)
    {
        switch (input = getchar())
        {
            case 'y':
            case 'Y':return 0;
            case 'n':
            case 'N':
            case EOF:return 1;
        }
    }
    
}

// put file descriptor 0 (i.e. stdin) into chr-by-chr mode
void set_cr_noecho_mode(){
    struct termios ttystate;
    tcgetattr(0,&ttystate);
    ttystate.c_lflag &= ~ICANON;
    ttystate.c_lflag &= ~ECHO;
    ttystate.c_cc[VMIN] = 1;
    tcsetattr(0,TCSANOW,&ttystate);
}

// how == 0 => save current mode, how == 1 => restore mode
int tty_mode(int how){
    static struct termios original_mode;
    if (how == 0)
    {
        tcgetattr(0,&original_mode);
    }
    else
    {
        return tcsetattr(0,TCSANOW,&original_mode);
    }
}

int main(int argc, char const *argv[])
{
    int response;
    tty_mode(0);
    set_cr_noecho_mode();
    response = get_response(QUESTION);
    tty_mode(1);
    return response;  
}

File: ch6/ch6_src/src/bin/play_again2.rs
use std::{error::Error, io::{self, Read}, mem, process::exit};

use libc::{tcgetattr, tcsetattr, termios, ECHO, ICANON, TCSANOW, VMIN};

const QUESTION:&str = "Do you want another translation";
fn main()->Result<(),Box<dyn Error>>{
    tty_mode(0);
    set_crmode();
    let response = get_response(QUESTION)?;
    tty_mode(1);
    exit(response)
}


fn tty_mode(how: i32)->i32{
    static mut ORIGINAL_MODE:termios = unsafe {mem::zeroed()};
    if how==0 {
        unsafe { tcgetattr(0, &raw mut ORIGINAL_MODE as *mut termios) }
    } else {
        unsafe { tcsetattr(0, TCSANOW, &raw const ORIGINAL_MODE as *const termios) }
    }
}

fn set_crmode(){
    let mut ttystate:termios = unsafe {mem::zeroed()};
    unsafe { tcgetattr(0, &raw mut ttystate as *mut termios) };
    ttystate.c_lflag &= !ICANON;
    ttystate.c_lflag &= !ECHO;
    ttystate.c_cc[VMIN] = 1;
    unsafe { tcsetattr(0, TCSANOW, &raw const ttystate as *const termios) };
}

fn getchar()->Result<char,Box<dyn Error>>{
    let mut buffer = [0;1];
    let mut stdin = io::stdin().lock();
    stdin.read_exact(&mut buffer)?;
    let char = buffer[0] as char;
    Ok(char)
}

fn get_response(question: &str) -> Result<i32,Box<dyn Error>> {
    println!("{} (y/n)?", question);

    loop {
        let input = getchar()?;

        match input {
            'y' | 'Y' => return Ok(0),
            _ => return Ok(1),
        }
    }
}